aea/agent_loop.py

```diff
@@ -0,0 +1,406 @@
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This module contains the implementation of an agent loop using asyncio."""
+import asyncio
+import datetime
+import logging
+import time
+from abc import ABC, abstractmethod
+from asyncio.events import AbstractEventLoop, TimerHandle
+from asyncio.futures import Future
+from asyncio.tasks import ALL_COMPLETED, FIRST_COMPLETED, Task
+from typing import Any, Callable, Dict, List, Optional, Set, Tuple, Union, cast
+
+from aea.exceptions import AEAException
+from aea.mail.base import InBox
+from aea.skills.base import Behaviour
+
+
+try:
+    from asyncio import create_task
+except ImportError:
+    # for python3.6!
+    from asyncio import ensure_future as create_task  # type: ignore
+
+
+logger = logging.getLogger(__file__)
+
+if False:  # MYPY compatible for types definitions
+    from aea.aea import AEA
+    from aea.agent import Agent
+
+
+def ensure_list(value: Any) -> List:
+    """Return [value] or value if value is a list."""
+    if not isinstance(value, list):
+        value = [value]
+    return value
+
+
+class AsyncState:
+    """Awaitable state."""
+
+    def __init__(self, initial_state: Any = None, loop: AbstractEventLoop = None):
+        """Init async state.
+
+        :param initial_state: state to set on start.
+        :param loop: optional asyncio event loop.
+        """
+        self._state = initial_state
+        self._watchers: Set[Future] = set()
+        self._loop = loop or asyncio.get_event_loop()
+
+    @property
+    def state(self) -> Any:
+        """Return current state."""
+        return self._state
+
+    @state.setter
+    def state(self, state: Any) -> None:
+        """Set state."""
+        if self._state == state:
+            return
+        self._state_changed(state)
+        self._state = state
+
+    def _state_changed(self, state: Any) -> None:
+        """Fulfill watchers for state."""
+        for watcher in list(self._watchers):
+            if state in watcher._states:  # type: ignore
+                self._loop.call_soon_threadsafe(
+                    watcher.set_result, (self._state, state)
+                )
+
+    async def wait(self, state_or_states: Union[Any, List[Any]]) -> Tuple[Any, Any]:
```

Comment:
consider using `Sequence` instead of `List` if it doesn't matter whether it's a tuple or a list

Reply:
done